oodeel/methods/dknn.py

```diff
@@ -56,54 +69,174 @@ def __init__(self, nearest: int = 50, use_gpu: bool = False):
+    def _prepare_layer_features(self, features: TensorType) -> np.ndarray:
         """
-        Constructs the index from ID data "fit_dataset", which will be used for
-        nearest neighbor search. Can operate on CPU or GPU based on the `use_gpu` flag.
+        Convert a feature tensor to a 2D numpy array and apply L2 normalization.
 
         Args:
-            fit_dataset: input dataset (ID) to construct the index with.
+            features (TensorType): Feature tensor to be processed.
+
+        Returns:
+            np.ndarray: Processed feature array with shape (num_samples, feature_dim)
+                and L2 normalized.
         """
-        fit_projected, _ = self.feature_extractor.predict(fit_dataset)
-        fit_projected = self.op.convert_to_numpy(fit_projected[0])
-        fit_projected = fit_projected.reshape(fit_projected.shape[0], -1)
-        norm_fit_projected = self._l2_normalization(fit_projected)
+        features_np = self.op.convert_to_numpy(features)
+        features_np = features_np.reshape(features_np.shape[0], -1)
+        return self._l2_normalization(features_np)
 
+    def _create_index(self, dim: int) -> faiss.IndexFlatL2:
+        """
+        Create a FAISS index for features of a given dimensionality.
+
+        Args:
+            dim (int): Dimensionality of the feature vectors.
+
+        Returns:
+            faiss.IndexFlatL2: A FAISS index instance, using GPU acceleration if
+                enabled.
+        """
         if self.use_gpu:
-            cpu_index = faiss.IndexFlatL2(norm_fit_projected.shape[1])
-            self.index = faiss.index_cpu_to_gpu(self.res, 0, cpu_index)
+            cpu_index = faiss.IndexFlatL2(dim)
+            return faiss.index_cpu_to_gpu(self.res, 0, cpu_index)
         else:
-            self.index = faiss.IndexFlatL2(norm_fit_projected.shape[1])
+            return faiss.IndexFlatL2(dim)
 
-        self.index.add(norm_fit_projected)
+    def _fit_layer(
+        self, layer_features: TensorType
+    ) -> Tuple[faiss.IndexFlatL2, Optional[np.ndarray]]:
+        """
+        Build a FAISS index for a single feature layer and compute initial scores for
+        aggregator fitting.
+
+        Args:
+            layer_features (TensorType): Feature tensor corresponding to a specific
+                layer.
+
+        Returns:
+            Tuple[faiss.IndexFlatL2, Optional[np.ndarray]]:
+                - The constructed FAISS index for the layer.
+                - Scores computed on a subset of features (first 1000 samples) if an
+                    aggregator is used; otherwise, None.
+        """
+        norm_features = self._prepare_layer_features(layer_features)
+        index = self._create_index(norm_features.shape[1])
+        index.add(norm_features)
+
+        scores = None
+        if self.aggregator is not None:
+            # Use only a subset of samples for computing initial scores, ensuring at
+            # least 2 neighbors (distance to self is 0)
```

Comment:
Why is it required for dknn? Maybe we could add a more descriptive comment?

Reply:
When fitting the aggregator, we use a subset of the training (fit) set to compute OOD scores. However, computing KNN scores with k=1 on the same samples used for fitting leads to zero-distance scores (since each sample is its own nearest neighbor), which is not meaningful for the aggregator.

To avoid this issue, we ensure that k >= 2 when computing KNN scores used to fit the aggregator, so that each sample’s score reflects a true distance to other points in the fit set.